### PR TITLE
Ra 1147 - Implement mappings management feature

### PR DESCRIPTION
**Pull request by pgutkowski**
This PR consist of following changes:
- add mapping table to concept view
- add mappingTable component to concept add/edit page, to handle adding and deleting mappings

Saving mappings won't work until this PR https://github.com/openmrs/openmrs-module-webservices.rest/pull/216 is merged

### CODE REVIEW

File: app/components/conceptAdd/conceptAdd.controller.js [hunk 1]
```diff
@@ -10,10 +10,12 @@
 import helpgif from '../../img/help.gif'
 
 ConceptAddController.$inject = 
-	['concept','serverLocales', 'loadClasses', 'loadDataTypes', 'conceptsService', '$location', '$routeParams', 'openmrsNotification'];
+	['concept','serverLocales', 'loadClasses', 'loadDataTypes', 'conceptsService',
+	'$location', '$routeParams', 'openmrsNotification', '$window'];
 	
 export default function ConceptAddController
-		(concept, serverLocales, loadClasses, loadDataTypes, conceptsService, $location, $routeParams, openmrsNotification){
+		(concept, serverLocales, loadClasses, loadDataTypes, conceptsService,
```
Comment (rkorytkowski) [inline app/components/conceptAdd/conceptAdd.controller.js]: Codacy Issue found: [This function has too many statements. (58)](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228898/issues/source?bid=3196969&fileBranchId=3354756#l17)

File: app/components/mappingsTable/mappingsTable.component.spec.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+            it("should divide mappings and store mappings to new reference terms at the end of array", function(){
+                component.mappings = mockMappings;
+                expect(component.mappings.indexOf(mockMapping1)).toEqualData(0);
+                component.divideMappings();
+                expect(component.mappings.indexOf(mockMapping1)).toEqualData(component.mappings.length-1);
+            })
+
+            it("mapping validation should return false if conceptSource uuid is not defined", function(){
+                var mock  = {conceptReferenceTerm: {conceptSource: {}}, conceptMapType: {uuid: "uuid"}, uuid: "uuid33" };
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.spec.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228924/issues/source?bid=3196969&fileBranchId=3354756#l90)

File: app/components/mappingsTable/mappingsTable.component.spec.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            expect(component.sources.length).toEqualData(1);
+            expect(component.mappings).toEqualData([]);
+        });
+
+        describe('component: mappingsTable, test functions', function(){
+            var mockMapping1, mockMapping2, mockMapping3, mockMappings, notificationMock;
+            notificationMock = testUtils.getNotificationMock();
+
+            beforeEach(function(){
+                mockMapping1 = {conceptReferenceTerm: {conceptSource: {uuid: "uuid"}, code: "code"}, conceptMapType: {uuid: "uuid"}, uuid: "uuid11" };
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.spec.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228924/issues/source?bid=3196969&fileBranchId=3354756#l59)

File: app/components/mappingsTable/mappingsTable.component.spec.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                mockMapping2 = {conceptReferenceTerm: {uuid: "uuid", conceptSource: {uuid: "uuid"}}, conceptMapType: {uuid: "uuid"}, uuid: "uuid22" };
+                mockMapping3 = {conceptReferenceTerm: {uuid: "uuid", conceptSource: {}}, conceptMapType: {uuid: "uuid"}, uuid: "uuid33" };
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.spec.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228924/issues/source?bid=3196969&fileBranchId=3354756#l61)

File: app/components/mappingsTable/mappingsTable.component.spec.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            expect(component.mappings).toEqualData([]);
+        });
+
+        describe('component: mappingsTable, test functions', function(){
+            var mockMapping1, mockMapping2, mockMapping3, mockMappings, notificationMock;
+            notificationMock = testUtils.getNotificationMock();
+
+            beforeEach(function(){
+                mockMapping1 = {conceptReferenceTerm: {conceptSource: {uuid: "uuid"}, code: "code"}, conceptMapType: {uuid: "uuid"}, uuid: "uuid11" };
+                mockMapping2 = {conceptReferenceTerm: {uuid: "uuid", conceptSource: {uuid: "uuid"}}, conceptMapType: {uuid: "uuid"}, uuid: "uuid22" };
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.spec.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228924/issues/source?bid=3196969&fileBranchId=3354756#l60)

File: app/components/mappingsTable/mappingsTable.component.js [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+mappingsTable.$inject = ['openmrsRest', 'openmrsNotification'];
+
+export default function mappingsTable(openmrsRest, openmrsNotification){
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: [This function has too many statements. (43)](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228923/issues/source?bid=3196969&fileBranchId=3354756#l12)

File: app/components/mappingsTable/mappingsTable.component.js [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                        break;
+                    }
+                }
+                if(angular.isUndefined(mapping.conceptReferenceTerm.uuid)) {
+                    mapping.conceptReferenceTerm.code = mapping.termQuery;
+                }
+            })
+        }
+    }
+}
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: ['mappingsTable' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228923/issues/source?bid=3196969&fileBranchId=3354756#l224)

File: app/components/mappingsTable/mappingsTable.component.js [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        if(map.conceptReferenceTerm.conceptSource.uuid === ""){
+            openmrsNotification.error("Invalid mapping: source is not specified");
+            return false;
+        }
+        //is already existing RefTerm defined?
+        var defRefTerm = angular.isDefined(map.conceptReferenceTerm.uuid);
+        //or else is new Term defined?
+        var defNewRefTerm = angular.isDefined(map.conceptReferenceTerm.conceptSource.uuid)
+                            &&angular.isDefined(map.conceptReferenceTerm.code);
+        if (defRefTerm||defNewRefTerm) return (defRefTerm||defNewRefTerm);
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228923/issues/source?bid=3196969&fileBranchId=3354756#l189)

File: app/components/conceptMappingsService/conceptMappings.service.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    return service;
+
+    function parseMappingSource(display){
+        return display.substring(0, display.lastIndexOf(":"));
+    }
+    function parseMappingCode(display){
+        //'+2' to cut off ': '
+        var nameIndex = display.search(termNameRegEx);
+        if(nameIndex === -1) nameIndex = display.length;
```
Comment (rkorytkowski) [inline app/components/conceptMappingsService/conceptMappings.service.js]: Codacy Issue found: [Expected '{' and instead saw 'nameIndex'.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228902/issues/source?bid=3196969&fileBranchId=3354756#l28)

File: app/components/conceptMappingsService/conceptMappings.service.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        return display.substring(display.lastIndexOf(": ")+2, nameIndex)
+    }
+    function parseMappingName(display){
+        var nameIndex = display.search(termNameRegEx);
+        if(nameIndex === -1) return "";
+        else return display.substring(nameIndex+1, display.length-1);
+    }
+}
```
Comment (rkorytkowski) [inline app/components/conceptMappingsService/conceptMappings.service.js]: Codacy Issue found: ['conceptMappingsService' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741228902/issues/source?bid=3196969&fileBranchId=3354756#l36)

File: app/components/mappingsTable/mappingsTable.component.js (hunk 6 above)
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: [This function has too many statements. (43)](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260425/issues/source?bid=3196969&fileBranchId=3354756#l12)

File: app/components/mappingsTable/mappingsTable.component.js (hunk 7 above)
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: ['mappingsTable' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260425/issues/source?bid=3196969&fileBranchId=3354756#l224)

File: app/components/mappingsTable/mappingsTable.component.js (hunk 8 above)
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260425/issues/source?bid=3196969&fileBranchId=3354756#l189)

File: app/components/conceptMappingsService/conceptMappings.service.js (hunk 9 above)
Comment (rkorytkowski) [inline app/components/conceptMappingsService/conceptMappings.service.js]: Codacy Issue found: [Expected '{' and instead saw 'nameIndex'.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260404/issues/source?bid=3196969&fileBranchId=3354756#l28)

File: app/components/conceptMappingsService/conceptMappings.service.js (hunk 10 above)
Comment (rkorytkowski) [inline app/components/conceptMappingsService/conceptMappings.service.js]: Codacy Issue found: ['conceptMappingsService' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260404/issues/source?bid=3196969&fileBranchId=3354756#l36)

File: app/components/conceptMappingsService/conceptMappings.service.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    function parseMappingCode(display){
+        //'+2' to cut off ': '
+        var nameIndex = display.search(termNameRegEx);
+        if(nameIndex === -1) nameIndex = display.length;
+        return display.substring(display.lastIndexOf(": ")+2, nameIndex)
+    }
+    function parseMappingName(display){
+        var nameIndex = display.search(termNameRegEx);
+        if(nameIndex === -1) return "";
```
Comment (rkorytkowski) [inline app/components/conceptMappingsService/conceptMappings.service.js]: Codacy Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260404/issues/source?bid=3196969&fileBranchId=3354756#l33)

File: app/components/conceptMappingsService/conceptMappings.service.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        else return display.substring(nameIndex+1, display.length-1);
```
Comment (rkorytkowski) [inline app/components/conceptMappingsService/conceptMappings.service.js]: Codacy Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260404/issues/source?bid=3196969&fileBranchId=3354756#l34)

File: app/components/conceptAdd/conceptAdd.controller.js (hunk 1 above)
Comment (rkorytkowski) [inline app/components/conceptAdd/conceptAdd.controller.js]: Codacy Issue found: [This function has too many statements. (58)](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260400/issues/source?bid=3196969&fileBranchId=3354756#l17)

File: app/components/mappingsTable/mappingsTable.component.spec.js (hunk 2 above)
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.spec.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260426/issues/source?bid=3196969&fileBranchId=3354756#l90)

File: app/components/mappingsTable/mappingsTable.component.spec.js (hunk 3 above)
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.spec.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260426/issues/source?bid=3196969&fileBranchId=3354756#l59)

File: app/components/conceptAdd/conceptAdd.controller.js (hunk 1 above)
Comment (rkorytkowski) [inline app/components/conceptAdd/conceptAdd.controller.js]: Codacy Issue found: [This function has too many statements. (58)](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741589643/issues/source?bid=3196969&fileBranchId=3354756#l17)

File: app/components/mappingsTable/mappingsTable.component.spec.js [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+            it("should divide mappings and store mappings to new reference terms at the end of array", function(){
+                component.mappings = mockMappings;
+                expect(component.mappings.indexOf(mockMapping1)).toEqualData(0);
+                component.divideMappings();
+                expect(component.mappings.indexOf(mockMapping1)).toEqualData(component.mappings.length-1);
+            })
+
+            it("mapping validation should return false if conceptSource uuid is not defined", function(){
+                var mock  = {conceptReferenceTerm: {conceptSource: {}}, conceptMapType: {uuid: "uuid"}, uuid: "uuid33" };
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.spec.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741589669/issues/source?bid=3196969&fileBranchId=3354756#l99)

File: app/components/conceptMappingsService/conceptMappings.service.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    function parseMappingName(display){
+        var nameIndex = display.search(termNameRegEx);
+        if(nameIndex === -1){
+            return "";
+        }
+        else{
+            return display.substring(nameIndex+1, display.length-1);
+        }
+    }
+}
```
Comment (rkorytkowski) [inline app/components/conceptMappingsService/conceptMappings.service.js]: Codacy Issue found: ['conceptMappingsService' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741589647/issues/source?bid=3196969&fileBranchId=3354756#l42)

File: app/components/mappingsTable/mappingsTable.component.js (hunk 6 above)
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: [This function has too many statements. (43)](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741589668/issues/source?bid=3196969&fileBranchId=3354756#l12)

File: app/components/mappingsTable/mappingsTable.component.js (hunk 7 above)
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: ['mappingsTable' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741589668/issues/source?bid=3196969&fileBranchId=3354756#l224)

File: app/components/mappingsTable/mappingsTable.component.js (hunk 8 above)
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js]: Codacy Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741589668/issues/source?bid=3196969&fileBranchId=3354756#l189)

File: app/components/mappingsTable/mappingsTable.component.js [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+mappingsTable.$inject = ['openmrsRest', 'openmrsNotification'];
+
+export default function mappingsTable(openmrsRest, openmrsNotification){
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js:12]: Codacy Issue found: [This function has too many statements. (43)](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741609016/issues/source?bid=3196969&fileBranchId=3354756#l12)

File: app/components/mappingsTable/mappingsTable.component.js [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+                        break;
+                    }
+                }
+                if(angular.isUndefined(mapping.conceptReferenceTerm.uuid)) {
+                    mapping.conceptReferenceTerm.code = mapping.termQuery;
+                }
+            })
+        }
+    }
+}
```
Comment (rkorytkowski) [inline app/components/mappingsTable/mappingsTable.component.js:226]: Codacy Issue found: ['mappingsTable' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741609016/issues/source?bid=3196969&fileBranchId=3354756#l226)

File: app/components/conceptMappingsService/conceptMappings.service.js (hunk 14 above)
Comment (rkorytkowski) [inline app/components/conceptMappingsService/conceptMappings.service.js:42]: Codacy Issue found: ['conceptMappingsService' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741608995/issues/source?bid=3196969&fileBranchId=3354756#l42)

File: app/components/conceptAdd/conceptAdd.controller.js (hunk 1 above)
Comment (rkorytkowski) [inline app/components/conceptAdd/conceptAdd.controller.js:17]: Codacy Issue found: [This function has too many statements. (58)](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741608991/issues/source?bid=3196969&fileBranchId=3354756#l17)

**Comment by pgutkowski** [comment]
II have fixed what could be fixed, I cannot shrink ConceptAddController or MappingsTableComponent, and I think moving statement with '$inject' after function body would make it less clear and readable

**Pull request merged**